### PR TITLE
Thread safety

### PR DESCRIPTION
**Pull request by atshaw43**
*Issue #, if available:*
It is possible to get a concurrent modification exception from adding PrecursorIds in two separate threads at the same time.

*Description of changes:*
Make the data structure for PrecursorIds thread safe.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/entities/EntityTest.java [hunk 1]
```diff
@@ -76,6 +76,18 @@ void testUnknownClassSerialization() {
...
+        Segment seg = new SegmentImpl(AWSXRay.getGlobalRecorder(), "test");
+        SubsegmentImpl subseg = new SubsegmentImpl(AWSXRay.getGlobalRecorder(), "test", seg);
+        subseg.addPrecursorId("myId1");
+        subseg.addPrecursorId("myId2");
+        String serializedSubSeg = subseg.serialize();
+
+        String expected = "\"precursor_ids\":[\"myId1\",\"myId2\"]";
+        assertThat(serializedSubSeg).contains(expected);
+    }
+
```
Comment (carolabadeer) [inline aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/entities/EntityTest.java:90]: Is there any way to mock the behaviour of adding precursor IDs in 2 threads and ensure that the new data structure handles that situation correctly?

File: aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/ConcurrencyTest.java [hunk 2]
```diff
@@ -64,4 +70,31 @@ void testManyThreads() throws InterruptedException {
...
+        Thread thread1 = new Thread(() -> {
+            while (true) {
+                subseg.addPrecursorId("ID" + new Random().nextInt());
+            }
+        });
+        thread1.start();
+
+        Callable<Void> callable = (Callable) () -> {
+            while (System.currentTimeMillis() - startTime < TimeUnit.SECONDS.toMillis(1)) {
+                Iterator it = subseg.getPrecursorIds().iterator();
```
Comment (jj22ee) [inline aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/ConcurrencyTest.java:89]: I think this `Callable` should also be using `addPrecursorId` so this function is tested being called in 2 threads.

Comment (atshaw43) [reply aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/ConcurrencyTest.java:89]: I was able to produce a concurrent modification exception with this code.
Iterating and changing the data structure is not allowed in a normal HashSet

If they are both adding, then it has to be the same bucket in the hash table at the same time which is harder to get to occur.

**Review by jj22ee — APPROVED** [review]
LGTM

**Pull request merged**